okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/UrlSelectorImpl.java

```diff
@@ -99,7 +99,13 @@ private static String switchWsToHttp(String url) {
     @Override
     public Optional<HttpUrl> redirectToNext(HttpUrl current) {
         int index = currentUrl.updateAndGet(operand -> (operand + 1) % baseUrls.size());
-        return redirectTo(current, baseUrls.get(index));
+        Optional<HttpUrl> newUrl = redirectTo(current, baseUrls.get(index));
+
+        if (newUrl.map(url -> url.equals(current)).orElse(false) && baseUrls.size() > 1) {
+            return redirectToNext(current);
```

Comment:
this potentially unbounded recursion is fishy. I understand that the underlying problem is that multiple callers may concurrently call redirectToNext and thus independently increment the currentUrl counter?

Reply:
yep this fix is awful (commit message is "bad fix"), I just wanted to see if it did fix it

yep exactly - so you end up with the retry url being the same as the one you just failed on. It's not easy to instead do it with thread locals because a call bounces around between some number of threads.

also wonder if we should instead just have a url selector impl per call, but maybe there was a reason we didn't do that to start with?

Reply:
it's tricky because you do want some information shared between calls:
```
client.foo(); // fails on server1, then tries server2 and succeeds
client.foo(); // should try server2 only
```
I think that's desired behavior? maybe round-robin is too naive an implementation here and instead we should keep per-url failure counts that age age off over time.

Reply:
pseudo code:
```
// select URL
map<url, failedCount> urls = ...;
set<url> goodUrls = urls.find(failedCount < 3);
return goodUrls.random();

// use URL
if (call(url) fails) {
  urls.get(url).incrementFailedCount();
}
```

Reply:
Counts don't seem necessary; if you just want to change default index on failure, all you have to do is to always redirect to the 'next' url after a failure (no matter what the current default is), and update default to what you redirected to. 

```
    @Override
    public Optional<HttpUrl> redirectToNext(HttpUrl current) {
        int index = (indexFor(current) + 1) % baseUrls.size();
        currentDefault.set(index)
        return redirectTo(current, baseUrls.get(index));
    }
```

OTOH I think there's an argument for rotating default urls on successful requests too (to avoid hot-spotting single node). In which case having fail counts makes more sense - you can track which nodes are failing more and send requests to the ones that aren't failing. But I think the above snippet solves this issue too.